### PR TITLE
[Banner] Update BannerTypicalUseExample with latest API changes

### PR DESCRIPTION
**Pull request by fluxminds**
Major changes:

- This PR updates the example for typical BannerView usage with latest BannerView API changes. 
- This PR add a separate example to each of three existing examples with no icon displayed on the banner view.
- This PR updated some constants in BannerTypicalUseExample. 

Screenshots Before Changes:
Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 11 26
Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 11 22

Screenshots After Changes:
Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 08 09
Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 08 06
Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 08 05

### CODE REVIEW

File: components/Banner/examples/supplemental/MDCBannerView.m [hunk 1]
```diff
@@ -18,7 +18,7 @@
 #import "MaterialTypography.h"
 
 typedef NS_ENUM(NSInteger, MDCBannerViewLayoutStyle) {
-  MDCBannerViewLayoutSingleLineStyle = 0,              // All elements lays on the same line
+  MDCBannerViewLayoutSingleLineStyle = 0,              // All elements on the same line
```
Comment (romoore) [inline components/Banner/examples/supplemental/MDCBannerView.m]: Please limit this PR to the examples. This file looks like it should be moved to a different PR.

**Review by romoore — DISMISSED** [review]
Please split out the source changes to a separate PR.

File: components/Banner/examples/supplemental/MDCBannerView.m (hunk 1 above)
Comment (fluxminds) [reply components/Banner/examples/supplemental/MDCBannerView.m]: Sorry. I made some code review change, and cherry-picked to both PR. The diff didn't display it correctly. I just rebased this branch, and the diff should now show correctly, which shouldn't include any source changes.

File: components/Banner/examples/resources/BannerAssets.xcassets/banner-email.imageset/Contents.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      "idiom" : "universal",
+      "filename" : "ic_email_black_3x_ios_24dp.png",
+      "scale" : "3x"
+    }
+  ],
+  "info" : {
+    "version" : 1,
+    "author" : "xcode"
+  }
+}
```
Comment (jverkoey) [inline components/Banner/examples/resources/BannerAssets.xcassets/banner-email.imageset/Contents.json:23]: If possible, please make use of an existing icon rather than adding a new asset to the examples.

**Comment by jverkoey** [comment]
Please add a before/after screenshot of the example to the PR description.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 3]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+  self.contentView.frame = self.view.bounds;
+  self.contentViewLabel.center = self.contentView.center;
+  CGRect exampleListTableViewFrame =
+      CGRectMake(0, self.view.bounds.size.height - exampleListTableViewHeight,
+                 self.view.bounds.size.width, exampleListTableViewHeight);
+  self.exampleListTableView.frame = exampleListTableViewFrame;
+
+
+  CGSize bannerViewSize =
+  [self.bannerView sizeThatFits:CGSizeMake(CGRectGetWidth(self.view.bounds), CGFLOAT_MAX)];
```
Comment (jverkoey) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m]: Indent this line.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 4]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+  bannerView.layoutMargins = margins;
+  [self.view addSubview:bannerView];
+  self.bannerView = bannerView;
+
+  MDCButton *button = bannerView.leadingButton;
+  [button setTitle:@"Dismiss" forState:UIControlStateNormal];
+  button.uppercaseTitle = YES;
   [button setTitleColor:self.colorScheme.primaryColor forState:UIControlStateNormal];
   button.backgroundColor = self.colorScheme.surfaceColor;
+  bannerView.trailingButton.hidden = YES;
```
Comment (jverkoey) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m:220]: This is somewhat surprising: is it expected that clients must hide the buttons they don't want to be shown?

I see two paths here:

1. Buttons are not hidden by default. We need to document that clients are expected to hide buttons that should not be shown.
2. Buttons are hidden by default. We need to document that clients are expected to unhide buttons if they want them to appear.

I'm not certain that either of these realities are particularly great, but either way we need to document this expected interaction with the component's buttons (not in this PR, but in a separate PR to the MDCBannerView public API).

**Review by jverkoey — APPROVED** [review]
LGTM though I have some concerns about the button visibility behavior.

File: components/Banner/examples/resources/BannerAssets.xcassets/banner-email.imageset/Contents.json (hunk 2 above)
Comment (fluxminds) [reply components/Banner/examples/resources/BannerAssets.xcassets/banner-email.imageset/Contents.json:23]: The other assets are included in the MDCCatalog/MDCDragon. This change was made to address a previous code review comment (https://github.com/material-components/material-components-ios/pull/6830#discussion_r264481238) to add assets to example assets folder. Not sure if there is a convention here?

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 3 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m]: Good catch. Addressed in 725ab5985288813e9e77af78e1dc1e437128da90

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 4 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:220]: I definitely agree. It is one of the few parts I feel unclear on the API, which is too implicit for developer to use. I will document it in the public API. https://github.com/material-components/material-components-ios/issues/7001 created to track the follow up work.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 5]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+
+  self.contentView.frame = self.view.bounds;
+  self.contentViewLabel.center = self.contentView.center;
+  CGRect exampleListTableViewFrame =
+      CGRectMake(0, self.view.bounds.size.height - exampleListTableViewHeight,
+                 self.view.bounds.size.width, exampleListTableViewHeight);
+  self.exampleListTableView.frame = exampleListTableViewFrame;
+
+  CGSize bannerViewSize =
+      [self.bannerView sizeThatFits:CGSizeMake(CGRectGetWidth(self.view.bounds), CGFLOAT_MAX)];
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m]: Why CGFloat_MAX?  Wouldn't you want the banner to fit within the view, or  is it placed within a scroll view?

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 6]
```diff
@@ -19,16 +19,14 @@
...
 #import "supplemental/MDCBannerView.h"
 
-static const CGFloat exampleListTableViewHeight = 300.0f;
-static const CGFloat exampleBannerContentMultiLineWidth = 320.0f;
-static const CGFloat exampleBannerContentSingleLineWidth = 350.0f;
+static const CGFloat exampleListTableViewHeight = 160.0f;
+static const CGFloat exampleBannerContentPadding = 10.0f;
 static NSString *const exampleShortText = @"tristique senectus et";
 static NSString *const exampleLongText =
-    @"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt.";
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m:27]: What does this string change have to do with the latest API changes?

**Review by romoore — COMMENTED** [review]
It looks like this PR is more than just updating an example for new APIs.  Please either correct the description or split the extra changes to a separate PR.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 7]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
   };
 }
 
+#pragma mark - UIViewController
+
+- (void)viewDidLayoutSubviews {
+  [super viewDidLayoutSubviews];
+
+  self.contentView.frame = self.view.bounds;
+  self.contentViewLabel.center = self.contentView.center;
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m]: This is incorrect because `center` is the center of the view's frame rectangle.

It should be:
```objc
self.contentViewLabel.center = CGPointMake(CGRectGetMidX(self.contentView.bounds), CGRectGetMidY(self.contentView.bounds));
```

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 8]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
   };
 }
 
+#pragma mark - UIViewController
+
+- (void)viewDidLayoutSubviews {
+  [super viewDidLayoutSubviews];
+
+  self.contentView.frame = self.view.bounds;
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m]: `contentView` already has an autoresizingMask set. You should not need to adjust its frame on each layout pass.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 9]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+  self.contentViewLabel.center = self.contentView.center;
+  CGRect exampleListTableViewFrame =
+      CGRectMake(0, self.view.bounds.size.height - exampleListTableViewHeight,
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m]: Please use `CGRectGet*()` methods to access bounds/frame rect properties.

```objc
CGRectMake(0, CGRectGetHeight(self.view.bounds) - exampleListTableViewHeight,
           CGRectGetWidth(self.view.bounds), exampleListTableViewHeight);
```

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 10]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+  CGSize bannerViewSize =
+      [self.bannerView sizeThatFits:CGSizeMake(CGRectGetWidth(self.view.bounds), CGFLOAT_MAX)];
+
+  // Adjust bannerViewContainer's frame
+  CGFloat topAreaInset = 0.0f;
+  if (@available(iOS 11.0, *)) {
+    topAreaInset = self.view.safeAreaInsets.top;
+  }
+  self.bannerView.frame =
+      CGRectMake(0.0f, topAreaInset, bannerViewSize.width, bannerViewSize.height);
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m:146]: You should probably constrain the banner view's width to be no greater than the view's bounds.  `sizeThatFits:` does not need to be constrained to the passed size.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 11]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+}
+
 #pragma mark - Internal helpers
 
 - (NSArray<BannerExampleUseInfo *> *)getBannerExampleList {
   NSMutableArray *bannerExampleList = [[NSMutableArray alloc] init];
   BannerExampleUseInfo *exampleUseInfo1 =
       [BannerExampleUseInfo infoWithIdentifier:@"example1"
-                                   displayName:@"Single Line"
+                                   displayName:@"Short Text with One Action"
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m:155]: These strings changes are not necessary for API changes. Please split them out.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 12]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+                                   displayName:@"Short Text with One Action (Icon)"
                               exampleUseTarget:self
-                            exampleUseSelector:@selector(showMultiLineAlignedButtonStyleBanner)];
+                            exampleUseSelector:@selector(showSingleLineStyleBannerWithIcon)];
   [bannerExampleList addObject:exampleUseInfo2];
 
   BannerExampleUseInfo *exampleUseInfo3 =
       [BannerExampleUseInfo infoWithIdentifier:@"example3"
-                                   displayName:@"MultiLine Stacked Button"
+                                   displayName:@"Long Text with Two Short Actions"
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m:169]: This isn't significantly longer text.  Consider doubling or trebling it.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 4 above)
Comment (romoore) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:220]: @wenyuzhang666 I thought we spoke and these properties would function similarly to UIButton. If they were never accessed, they would not be created. If they were accessed, they would be created dynamically.  This line looks like it's purpose is the side-effect of creating the button.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 13]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
-                                  self.bannerViewContainer.frame.size.height / 2);
+
+  CGSize bannerViewSize =
+      [bannerView sizeThatFits:CGSizeMake(CGRectGetWidth(self.view.bounds), CGFLOAT_MAX)];
+  bannerView.frame = CGRectMake(0.0f, topAreaInset, bannerViewSize.width, bannerViewSize.height);
+}
+
+- (void)showSingleLineStyleBannerWithIcon {
+  [self showSingleLineStyleBanner];
+  // TODO: https://github.com/material-components/material-components-ios/issues/6838
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m]: Please add a comment describing what the TODO is. The link is helpful to track the work, but there should be an explanation of the bug/issue/clean-up as well.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m [hunk 14]
```diff
@@ -129,230 +130,238 @@ + (NSDictionary *)catalogMetadata {
...
+                       action:@selector(dismissBanner)
              forControlEvents:UIControlEventTouchUpInside];
 
-  self.bannerView = bannerView;
-  [self.bannerViewContainer addSubview:bannerView];
-  CGSize bannerViewSize =
-      [bannerView sizeThatFits:CGSizeMake(exampleBannerContentMultiLineWidth, CGFLOAT_MAX)];
-  bannerView.frame = CGRectMake(0.0f, 0.0f, bannerViewSize.width, bannerViewSize.height);
-
   // Adjust bannerViewContainer's frame
```
Comment (romoore) [inline components/Banner/examples/BannerTypicalUseExampleViewController.m]: Why do you have to adjust the banner's frame? Isn't this handled in `viewDidLayoutSubviews`?

File: components/Banner/examples/resources/BannerAssets.xcassets/banner-email.imageset/Contents.json (hunk 2 above)
Comment (romoore) [reply components/Banner/examples/resources/BannerAssets.xcassets/banner-email.imageset/Contents.json:23]: Agreed - the catalog assets aren't available to the examples targets unless linked against the catalog.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 4 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:220]: @romoore I remember the conversation. I double checked the behavior of `UIButton` after that,  UIButton's `imageView` is always created and set hidden by default if no image is set, which led to the non-lazy load approach right now. I can go with the lazy load one too.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 13 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m]: Sure. This TODO was done. Addressed in https://github.com/material-components/material-components-ios/pull/6994/commits/808dd2535cf1f193b23024da6c1c20b98092bed9.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 12 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:169]: If it is doubled, the text will be trimmed on the screen since the limit of number of lines is 3.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 9 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m]: Good point. I missed this one. Addressed in https://github.com/material-components/material-components-ios/pull/6994/commits/808dd2535cf1f193b23024da6c1c20b98092bed9.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 8 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m]: Addressed in https://github.com/material-components/material-components-ios/pull/6994/commits/808dd2535cf1f193b23024da6c1c20b98092bed9.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 7 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m]: Addressed in https://github.com/material-components/material-components-ios/pull/6994/commits/808dd2535cf1f193b23024da6c1c20b98092bed9.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 5 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m]: Addressed in ccab2788839d0a28c0b4925a667ddf699304cdfe. This one is not placed in a scroll view.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 14 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m]: Good point! It was necessary before because frame needs to be set before `viewDidLayoutSubviews`. Some changes made in https://github.com/material-components/material-components-ios/pull/6994/commits/3972bb8b4d40b8882b6275e31863f5da28be00f6 to address it and removed some duplicate code blocks. It's cleaner now.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 6 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:27]: Updated the description to reflect the constant change here.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 11 above)
Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:155]: Updated the description to reflect the constant change here.

File: components/Banner/examples/BannerTypicalUseExampleViewController.m (hunk 12 above)
Comment (romoore) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:169]: That sounds like a good example to show.

Comment (fluxminds) [reply components/Banner/examples/BannerTypicalUseExampleViewController.m:169]: Sure. I will add it. Tracked issue https://github.com/material-components/material-components-ios/issues/7017.
